### ISSUE TITLE
Predefining Node IP-Network

### ISSUE THREAD

**Issue by Floppy012** [vitobotta/hetzner-k3s#169]
I would love to use this project, but I'm having an annoying problem. I wonder what could be the smartest way to solve this:

**My scenario**

I have 3 Hetzner Cloud networks (Management, Development, Production). Those three networks to separate the different environments for security reasons. They are connected through a pfSense instance which is connected to all three networks and does the routing/firewall.

The networks are `10.0/16`, `10.10/16` and `10.20/16`. And then they are subnetted again into different sections (to keep them tidy). In each network I have some servers that have to be there (e.g. database nodes, the pfsense itself).

I want to spin up 2 clusters (one in the Development and one in the Production network). However, I don't want the nodes to just randomly get IP-Adresses from the whole /16 range they're in. They should all be contained in their own subnet (again, to keep things tidy).

So to clarify:

| Server | IP-Address |
| ------- | ----------------- |
| pfSense | 10.20.0.2 |
| database | 10.20.10.1 |
| k3s-node1 | 10.20.100.1 |
| k3s-node2 | 10.20.100.2 |
| ... | ... |
| k3s-node25 | 10.20.100.25 |

I cannot put the clusters in their own networks, as I can not connect the router to more than three networks and there is some connectivity between the networks required.

So the only way to achieve this would be to handle it with the creation of the servers. That itself is a shitshow as the Hetzner API doesn't allow to pass just a range. If I didn't miss anything I would have to create a server without a private network attached, then attach it afterwards through the API to the private network with an IP-Address from the desired range that also isn't already occupied. Which would mean to extract every single IP address of every server in the network, then checking if the IP is in the range, build a list and then determine what the next free IP-Address is in that range and use that to attach the server to the network.

It feels like a lot of unnecessary work that has to be done. Additionally it would have to be done in this tool and the autoscaler, as the autoscaler also doesn't support such a feature because Hetzner doesn't support it.

The other way would be to ditch any effort of keeping the networks themselves tidy and just letting the network do its thing. 

Maybe there is another way that I currently don't see. But I'm feeling like I'm standing with my back against a wall in the current situation.

**Comment by vitobotta** [comment]
I am not sure if I understand the problem correctly. The tool allows you to create a cluster specifying an existing network instead, so you can just set the name of the network you want to use in the config file. Won't this work for you?

**Comment by Floppy012** [comment]
Yes. That will work. But it will randomly assign IP-Addresses to the nodes. If you have your Network separated into different subnets, then some servers will get IP-Addresses from one subnet, then another subnet. It doesn't matter on a technical layer but when trying to keep things in order it gets annoying.


Here is an example of what I mean by subnetting a network:
![image](https://user-images.githubusercontent.com/7030787/224509918-4ed9ca13-cdc3-4932-a2c5-6eb9a4325bed.png)

**Comment by vitobotta** [comment]
So ..what are you suggesting/needing from this tool?

**Comment by Floppy012** [comment]
It was a question to you and people using this tool. Maybe someone has an idea I haven't thought of

**Comment by vitobotta** [comment]
> Yes. That will work. But it will randomly assign IP-Addresses to the nodes. If you have your Network separated into different subnets, then some servers will get IP-Addresses from one subnet, then another subnet. It doesn't matter on a technical layer but when trying to keep things in order it gets annoying.
> 
> Here is an example of what I mean by subnetting a network: ![image](https://user-images.githubusercontent.com/7030787/224509918-4ed9ca13-cdc3-4932-a2c5-6eb9a4325bed.png)

I am not particularly good with networks but you can specify the `private_network_subnet` in the config. Is that not helping?

**Comment by Floppy012** [comment]
> I am not particularly good with networks but you can specify the `private_network_subnet` in the config. Is that not helping?

Sadly not. The cloud network seems to assign random IP-Addresses from any defined subnet in the network. So some will be in the 10.0.0.0/24 some in the 10.0.1.0/24 and some in the 10.0.2.0/24.

**Comment by vitobotta** [comment]
Afraid there is not much I can do about it unfortunately if the API doesn't allow it.

**Comment by Floppy012** [comment]
> Afraid there is not much I can do about it unfortunately if the API doesn't allow it.

Alright. Had a little hope that I've overseen something that would make this possible. But now that four eyes have looked at it and don't see any way to accomplish this I just have to accept it and put all in one subnet.

Thanks for spending your time looking at it 😃
